File: algorithms/2_medium/E/01921_eliminate_maximum_number_of_monsters/solution.cpp
class Solution {
public:
    static int eliminateMaximum(const vector<int>& dist, const vector<int>& speed) {
        assert(dist.size() == speed.size());
        const size_t len = dist.size();

        auto arr = vector<int>();
        arr.reserve(len);
        for (size_t i = 0; i < len; i += 1) {
            assert(speed[i] > 0);
            arr.push_back((dist[i] + speed[i] - 1) / speed[i]);
        }
        ranges::sort(arr);

        for (size_t i = 0; i < len; i += 1) {
            if (static_cast<size_t>(arr[i]) < i + 1) {
                return static_cast<int>(i);
            }
        }

        return static_cast<int>(len);
    }
};

File: algorithms/2_medium/E/01921_eliminate_maximum_number_of_monsters/solution.rs
impl Solution {
    pub fn eliminate_maximum(dist: Vec<i32>, speed: Vec<i32>) -> i32 {
        assert!(dist.len() == speed.len());
        let len = dist.len();

        let mut arr: Vec<i32> = Vec::with_capacity(len);
        for i in 0..len {
            assert!(speed[i] > 0);
            arr.push((dist[i] + speed[i] - 1) / speed[i]);
        }
        arr.sort_unstable();

        for i in 0..len {
            if (arr[i] as usize) < i + 1 {
                return i as i32;
            }
        }

        len as i32
    }
}
